test/e2e/performanceprofile/functests/1_performance/cpu_management.go

```diff
@@ -889,8 +918,154 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
+			Expect(err).ToNot(HaveOccurred(), "Best-effort pod did not become ready in time")
+			testlog.Infof("BestEffort pod %s/%s was successfully created", bestEffortPod.Namespace, bestEffortPod.Name)
+
+			By("Geting Information for guaranteed POD containers")
+			GuPods := getConfigJsonInfo(guaranteedPod, "test", workerRTNode)
+			for _, pod := range GuPods {
+				if pod.Annotations.ContainerName == "test" {
+					guaranteedPodCpus, err = cpuset.Parse(string(pod.Linux.Resources.CPU.CPUs))
+				} else if pod.Annotations.ContainerName == "POD" {
+					guaranteedInitPodCpus, err = cpuset.Parse(string(pod.Linux.Resources.CPU.CPUs))
+				}
+				Expect(err).ToNot(HaveOccurred(), "Failed to parse GU POD cpus")
+			}
+
+			By("Geting Information for BestEffort POD containers")
+			BEPods := getConfigJsonInfo(bestEffortPod, "test", workerRTNode)
+			for _, pod := range BEPods {
+				if pod.Annotations.ContainerName == "test" {
+					bestEffortPodCpus, err = cpuset.Parse(string(pod.Linux.Resources.CPU.CPUs))
+				} else if pod.Annotations.ContainerName == "POD" {
+					bestEffortInitPodCpus, err = cpuset.Parse(string(pod.Linux.Resources.CPU.CPUs))
+				}
+				Expect(err).ToNot(HaveOccurred(), "Failed to parse BE POD cpus")
+			}
+
+			By("Validating CPU allocation for Guaranteed and Best-Effort pod containers")
+			isolatedCpus, err := cpuset.Parse(string(*profile.Spec.CPU.Isolated))
+			Expect(err).ToNot(HaveOccurred(), "Failed to parse isolated CPU set from performance profile")
+			reservedCpus, err := cpuset.Parse(string(*profile.Spec.CPU.Reserved))
+			Expect(err).ToNot(HaveOccurred(), "Failed to parse reserved CPU set from performance profile")
+
+			Expect(guaranteedInitPodCpus.IsSubsetOf(reservedCpus)).
+				To(BeTrue(), "Guaranteed Init pod CPUs (%s) are not strictly within the reserved set (%s)", guaranteedInitPodCpus, reservedCpus)
+			Expect(guaranteedPodCpus.IsSubsetOf(isolatedCpus)).
+				To(BeTrue(), "Guaranteed pod CPUs (%s) are not strictly within the isolated set (%s)", guaranteedPodCpus, isolatedCpus)
+
+			availableForBestEffort := isolatedCpus.Union(reservedCpus).Difference(guaranteedPodCpus)
+			Expect(bestEffortInitPodCpus.IsSubsetOf(reservedCpus)).
+				To(BeTrue(), "Best-Effort Init pod CPUs (%s) include CPUs not allowed (%s)", bestEffortInitPodCpus, availableForBestEffort)
+			Expect(bestEffortPodCpus.IsSubsetOf(availableForBestEffort)).
+				To(BeTrue(), "Best-Effort pod CPUs (%s) include CPUs not allowed (%s)", bestEffortPodCpus, availableForBestEffort)
+		})
+	})
+
 })
 
+func extractConfigInfo(output string) (*ContainerConfig, error) {
+	var config ContainerConfig
+	output = strings.TrimSpace(output)
+	err := json.Unmarshal([]byte(output), &config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config.json: %v", err)
+	}
+	return &config, nil
+}
+
+func getConfigJsonInfo(pod *corev1.Pod, containerName string, workerRTNode *corev1.Node) []*ContainerConfig {
```

Comment:
> One approach to testing is to verify whether `runc`/`crun` isolates itself from the isolated CPUs by following the method defined in the OCI specification, which involves generating a `config.json` and adhering to the specified CPUs based on the applied profile.

Well, does the process which writes `config.json` run only on reserved CPUs? And does `crun` (or `runc`) honor the settings on `config.json` when setting its own affinity? IOW, who decides and who enforces that `runc` (or `crun`) only runs on reserved CPUs?

your test seems fine for workload, but there are open questions for the infra proper

> Alternatively, a tracing tool like `bpftrace` can be used to monitor all `runc` calls and inspect their CPU affinity. However, this method is quite invasive and may be challenging to implement IMO.

I agree, but the problem here is first and foremost testing the right thing.

Reply:
@ffromani Do you have any other way in mind?

Reply:
To ensure the check is robust, I have added another `Expect` to verify whether it honors the provided CPUs through the profile. You can review the change here: [Line](https://github.com/openshift/cluster-node-tuning-operator/blob/143f8310134136760a5082c520e060780ac1f869/test/e2e/performanceprofile/functests/1_performance/cpu_management.go#L993) The latest commit addresses your concerns.

Reply:
@SargunNarula would it be correct to generalize the problem as whether host's processes are respecting the isolated cpus, when they are allocated exclusively for a Guaranteed pod? 

Another suggestion I had in mind is to examine if we can add a test for that on crun/runc repo.
Maybe the test's context there would be more suitable for this kind of verification.

Reply:
@Tal-or Yes, the problem statement is correct. However, I cannot definitively state which other host processes we might consider for this purpose.

> Create a short shell script that runs and prints its associated CPUs

The concept of using a shell script was addressed in the original verification process, where we employed a wrapper around the runc process to capture and log its associated CPUs to a separate file. The challenge, however, was that despite replacing the binary with the wrapper, capturing the CPUs remained difficult due to the extremely short-lived nature of the runc process. The original test logic was designed to address that testing approach, as detailed in this [link](https://github.com/SargunNarula/cluster-node-tuning-operator/blob/b4dfe4a03da80f0d9ec1131cbd958edca9514f8a/test/e2e/performanceprofile/functests/1_performance/cpu_isolation.go).

> Add a test for that in the crun/runc repository

For `runc`, it might be feasible to design a test. However, verifying its behavior with CPU ranges defined by the performance profile presents a challenge, where this test could be valuable.

For `crun`, this test would not be applicable due to its distinct workflow. Unlike `runc`, `crun` does not rely on init containers. Instead, its workflow involves the following steps:
- Creating a temporary bind mount of the current executable.
- Remounting it as read-only.
- Opening the remounted file.
- Unmounting and deleting the temporary file.
- Using the file descriptor to execute the program.

This difference in implementation makes this testing approach unsuitable for crun.

Reply:
>the challenge, however, was that despite replacing the binary with the wrapper, capturing the CPUs remained difficult due to the extremely short-lived nature of the runc process. 

If we're creating our own shell script for test purposes we can make it as longer as needed and the script will print the CPUs. The script doesn't have to be related to `runc`/`crun` we only want to make sure it doesn't violate the isolated cpus isolation. 

> as detailed in this [link](https://github.com/SargunNarula/cluster-node-tuning-operator/blob/b4dfe4a03da80f0d9ec1131cbd958edca9514f8a/test/e2e/performanceprofile/functests/1_performance/cpu_isolation.go). 

The link is a go file, not sure if that's what you meant.

>For runc, it might be feasible to design a test. However, verifying its behavior with CPU ranges defined by the performance profile presents a challenge, where this test could be valuable.

>For crun, this test would not be applicable due to its distinct workflow. Unlike runc, crun does not rely on init containers. Instead, its workflow involves the following steps:

>Creating a temporary bind mount of the current executable.
Remounting it as read-only.
Opening the remounted file.
Unmounting and deleting the temporary file.
Using the file descriptor to execute the program.
This difference in implementation makes this testing approach unsuitable for crun.

Ok too complex, I get it.

Reply:
If we aim to check the CPU usage of the shell script, we could extend its runtime. However, this would deviate from the primary objective, which is to verify whether the `runc` process is violating CPU isolation. 

In the future, this might need to be extended to other processes, but as of now, that is not a requirement. The original bug specifically pertains to the `runc` process utilizing isolated CPUs. [Bug link](https://bugzilla.redhat.com/show_bug.cgi?id=1910386)

The Go file contains a single test for this bug, utilizing the shell script approach. [Link](https://github.com/SargunNarula/cluster-node-tuning-operator/blob/b4dfe4a03da80f0d9ec1131cbd958edca9514f8a/test/e2e/performanceprofile/functests/1_performance/cpu_isolation.go)